src/ILCompiler.ReadyToRun/src/ObjectWriter/SectionBuilder.cs

```diff
@@ -0,0 +1,921 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
+using System.Reflection.PortableExecutable;
+using System.Runtime.CompilerServices;
+
+namespace ILCompiler.PEWriter
+{
+    public static class RelocType
+    {
+        /// <summary>
+        /// No relocation required
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_ABSOLUTE = 0x00;
+
+        /// <summary>
+        /// The 32-bit address without an image base (RVA)
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_ADDR32NB = 0x02;
+
+        /// <summary>
+        /// 32 bit address base
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_HIGHLOW = 0x03;
+
+        /// <summary>
+        /// Thumb2: based MOVW/MOVT
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_THUMB_MOV32 = 0x07;
+
+        /// <summary>
+        /// 64 bit address base
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_DIR64 = 0x0A;
+
+        /// <summary>
+        /// 32-bit relative address from byte following reloc
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_REL32 = 0x10;
+
+        /// <summary>
+        /// Thumb2: based B, BL
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_THUMB_BRANCH24 = 0x13;
+
+        /// <summary>
+        /// Arm64: B, BL
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_ARM64_BRANCH26 = 0x14;
+
+        /// <summary>
+        /// 32-bit relative address from byte starting reloc
+        /// This is a special NGEN-specific relocation type
+        /// for relative pointer (used to make NGen relocation
+        /// section smaller)
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_RELPTR32 = 0x7C;
+
+        /// <summary>
+        /// 32 bit offset from base of section containing target
+        /// </summary>
+        public const ushort IMAGE_REL_SECREL = 0x80;
+
+        /// <summary>
+        /// ADRP
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_ARM64_PAGEBASE_REL21 = 0x81;
+
+        /// <summary>
+        /// ADD/ADDS (immediate) with zero shift, for page offset
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_ARM64_PAGEOFFSET_12A = 0x82;
+
+        /// <summary>
+        /// LDR (indexed, unsigned immediate), for page offset
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_ARM64_PAGEOFFSET_12L = 0x83;
+    }
+
+    /// <summary>
+    /// Opaque key representing a single block.
+    /// </summary>
+    public struct BlockHandle : IEquatable<BlockHandle>, IEqualityComparer<BlockHandle>
```

Comment:
Why do these implement `IEqualityComparer`?

Reply:
I was under the impression that I need an IEqualityComparer when I want to use BlockHandle as a key in a Dictionary as happens in SessionBuilder._blockDictionary.

Reply:
Hmm, how can Directory hash the keys in order to facilitate fast lookup without IEqualityComparer?

Reply:
sorry, Dictionary

Reply:
This code is not passing an IEqualityComparer to the constructor of the dictionary.

Dictionary will call `Object.GetHashCode`:

https://github.com/dotnet/corert/blob/d966f7361ca8d295bec58af7853c74ebe03f2b47/src/System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs#L377-L380

Reply:
Thanks for explanation. One way or another, I have removed this struct altogether in the 3rd commit based on your suggestion :-).